base/expr.jl

```diff
@@ -535,71 +535,131 @@ non-termination if the method calls some other method that does not terminate.
-and is a convenient shortcut.
+- `:notaskstate`
 
-!!! note
-    This setting is particularly useful since it allows the compiler to evaluate a call of
-    the applied method when all the call arguments are fully known to be constant, no matter
-    if the call results in an error or not.
-
-    `@assume_effects :total_may_throw` is similar to [`@pure`](@ref) with the primary
-    distinction that the `:consistent`-cy requirement applies world-age wise rather
-    than globally as described above. However, in particular, a method annotated
-    `@pure` should always be `:total` or `:total_may_throw`.
-    Another advantage is that effects introduced by `@assume_effects` are propagated to
-    callers interprocedurally while a purity defined by `@pure` is not.
+!!! warning
+    `:total` is a very strong assertion and will likely gain additional semantics
+    in future versions of julia (e.g. if additional effects are added and included
+    in the definition of `:total`). As a result, it should be used with care.
+    Whenever possible, prefer to use the minimum possible set of specific effect
+    assertions required for a particular application. In cases where a large
+    number of effect overrides apply to a set of functions, a custom macro is
+    recommended over the use of `:total`.
+
+## Negated effects
+
+Effect names may be prefixed by `!` to indicate that the effect should be removed
+from an earlier meta effect. For example, `:pure !:nothrow` indicates that while
+the call is generally pure, it may however throw.
+
+## Comparison to @pure
+`@assume_effects :consistent :effect_free :terminates_globally` is similar to [`@pure`](@ref) with the primary
+distinction that the `:consistent`-cy requirement applies world-age wise rather
+than globally as described above. However, in particular, a method annotated
+`@pure` should always be at least `:consistent :effect_free :terminates_globally`.
+Another advantage is that effects introduced by `@assume_effects` are propagated to
+callers interprocedurally while a purity defined by `@pure` is not.
 """
 macro assume_effects(args...)
-    (consistent, effect_free, nothrow, terminates_globally, terminates_locally) =
-        (false, false, false, false, false, false)
+    (consistent, effect_free, nothrow, terminates_globally, terminates_locally, notaskstate) =
+        (false, false, false, false, false, false, false)
     for setting in args[1:end-1]
-        if isa(setting, QuoteNode)
-            setting = setting.value
+        val = true
+        org_setting = setting
+        while true
+            if isa(setting, Expr) && setting.head === :call
+                if setting.args[1] == :(!)
+                    val = !val
+                    setting = setting.args[2]
+                end
+            elseif isa(setting, QuoteNode) && isa(setting.value, Symbol)
+                setting = setting.value
+            else
+                break
+            end
         end
```

Comment:
```suggestion
        org_setting = setting
        setting, val = compute_setting(setting)
```
with
```julia
function compute_setting(@nospecialize(setting), val::Bool=true)
    if isexpr(setting, :call) && settings.args[1] === :(!)
        return compute_setting(setting.args[2], !val)
    elseif isa(setting, QuoteNode)
        return compute_setting(setting.value, val)
    else
        return (setting, val)
    end
end
```
?

Reply:
Sure. I'm assuming nobody will build exprs big enough that this bites us :)